lib/widgets/recent_dm_conversations.dart

```diff
@@ -0,0 +1,119 @@
+import '../model/recent_dm_conversations.dart';
+import 'content.dart';
+import 'icons.dart';
+import 'message_list.dart';
+import 'page.dart';
+import 'store.dart';
+import 'text.dart';
+
+class RecentDmConversationsPage extends StatefulWidget {
+  const RecentDmConversationsPage({super.key});
+
+  static Route<void> buildRoute({required BuildContext context}) {
+    return MaterialAccountPageRoute(context: context,
+      builder: (context) => const RecentDmConversationsPage());
+  }
+
+  @override
+  State<RecentDmConversationsPage> createState() => _RecentDmConversationsPageState();
+}
+
+class _RecentDmConversationsPageState extends State<RecentDmConversationsPage> with PerAccountStoreAwareStateMixin<RecentDmConversationsPage> {
+  RecentDmConversationsView? model;
+
+  @override
+  void onNewStore() {
+    model?.removeListener(_modelChanged);
+    model = PerAccountStoreWidget.of(context).recentDmConversationsView
+      ..addListener(_modelChanged);
+  }
+
+  void _modelChanged() {
+    setState(() {
+      // The actual state lives in [model].
+      // This method was called because that just changed.
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final sorted = model!.sorted;
+    return Scaffold(
+      appBar: AppBar(title: const Text('Direct messages')),
+      body: ListView.builder(
+        itemCount: sorted.length,
+        itemBuilder: (context, index) => RecentDmConversationsItem(narrow: sorted[index])));
+  }
+}
+
+class RecentDmConversationsItem extends StatelessWidget {
+  const RecentDmConversationsItem({super.key, required this.narrow});
+
+  final DmNarrow narrow;
+
+  final _avatarSize = 32.0;
+  final _avatarBorderRadius = 3.0;
+
+  @override
+  Widget build(BuildContext context) {
+    final store = PerAccountStoreWidget.of(context);
+    final selfUser = store.users[store.account.userId]!;
+
+    final String title;
+    final Widget avatar;
+    switch (narrow.otherRecipientIds) {
+      case []:
+        title = selfUser.fullName;
+        avatar = Avatar(userId: selfUser.userId, size: _avatarSize, borderRadius: _avatarBorderRadius);
+      case [var otherUserId]:
+        final otherUser = store.users[otherUserId];
+        title = otherUser?.fullName ?? '(unknown user)';
+        avatar = otherUser != null
+          ? Avatar(userId: otherUser.userId, size: _avatarSize, borderRadius: _avatarBorderRadius)
+          : SizedBox.square(dimension: _avatarSize);
+      default:
+        // TODO(i18n): List formatting, like you can do in JavaScript:
+        //   new Intl.ListFormat('ja').format(['Chris', 'Greg', 'Alya'])
+        //   // 'Chris、Greg、Alya'
+        title = narrow.otherRecipientIds.map((id) => store.users[id]!.fullName).join(', ');
+        avatar = SizedBox.square(
+          dimension: _avatarSize,
+          child: ClipRRect(
+            borderRadius: BorderRadius.all(Radius.circular(_avatarBorderRadius)),
+            clipBehavior: Clip.antiAlias,
```

Comment:
Hmm yeah, I see, this is a bit annoying.

Still I think this is better than having `Avatar` sometimes take responsibility for its shape, and sometimes not. That feels like having that widget live at two different conceptual layers depending on the arguments it's constructed with.

I think the Flutter-upstream-style solution here would probably be to decompose `Avatar` into two widgets:
* one named like `AvatarImage` that just builds the RealmContentNetworkImage (consulting the store in order to do so);
* one named like `AvatarShape` that is very much like the `_AvatarWrapper` in your previous revision, but taking size and borderRadius as parameters;
* and then also perhaps one named `Avatar` that is just a convenience composing those two.

Then this switch would compute a `Widget avatar` that's either an AvatarImage or an Icon, and then in a single spot afterward this build method would enclose that in an AvatarShape.

Reply:
That makes sense! I can go for that in my next revision.